cdm/core/src/main/java/ucar/nc2/dataset/VariableDS.java

```diff
@@ -609,172 +612,211 @@ public void removeCoordinateSystem(CoordinateSystem cs) {
-    return scaleMissingUnsignedProxy.getMissingValues();
+    return convertMissing != null ? convertMissing.getMissingValues() : new double[] {0};
   }
 
   @Override
   public boolean isMissingValue(double val) {
-    return scaleMissingUnsignedProxy.isMissingValue(val);
+    return convertMissing != null ? convertMissing.isMissingValue(val) : false;
   }
 
   /** @deprecated Use NetcdfDataset.builder() */
   @Deprecated
   @Override
   public void setFillValueIsMissing(boolean b) {
-    scaleMissingUnsignedProxy.setFillValueIsMissing(b);
+    if (convertMissing != null) {
+      convertMissing.setFillValueIsMissing(b);
+    }
   }
 
   /** @deprecated Use NetcdfDataset.builder() */
   @Deprecated
   @Override
   public void setInvalidDataIsMissing(boolean b) {
-    scaleMissingUnsignedProxy.setInvalidDataIsMissing(b);
+    if (convertMissing != null) {
+      convertMissing.setInvalidDataIsMissing(b);
+    }
+  }
+
+  public boolean missingDataIsMissing() {
+    return builder().missingDataIsMissing;
+  }
+
+  public boolean fillValueIsMissing() {
+    return builder().fillValueIsMissing;
+  }
+
+  public boolean invalidDataIsMissing() {
+    return builder().invalidDataIsMissing;
   }
 
   /** @deprecated Use NetcdfDataset.builder() */
   @Deprecated
   @Override
   public void setMissingDataIsMissing(boolean b) {
-    scaleMissingUnsignedProxy.setMissingDataIsMissing(b);
+    if (convertMissing != null) {
+      convertMissing.setMissingDataIsMissing(b);
+    }
   }
 
   @Nullable
   @Override
   public DataType getScaledOffsetType() {
-    return scaleMissingUnsignedProxy.getScaledOffsetType();
+    return scaleOffset != null ? scaleOffset.getScaledOffsetType() : dataType;
   }
 
   @Override
   public DataType getUnsignedConversionType() {
-    return scaleMissingUnsignedProxy.getUnsignedConversionType();
+    return unsignedConversion != null ? unsignedConversion.getOutType() : dataType;
   }
 
   @Override
   public DataType.Signedness getSignedness() {
-    return scaleMissingUnsignedProxy.getSignedness();
+    return unsignedConversion != null ? unsignedConversion.getSignedness() : dataType.getSignedness();
   }
 
   @Override
   public double applyScaleOffset(Number value) {
-    return scaleMissingUnsignedProxy.applyScaleOffset(value);
+    return scaleOffset != null ? scaleOffset.removeScaleOffset(value) : value.doubleValue();
   }
 
   @Override
   public Array applyScaleOffset(Array data) {
-    return scaleMissingUnsignedProxy.applyScaleOffset(data);
+    return scaleOffset != null ? scaleOffset.removeScaleOffset(data) : data;
   }
 
   @Override
   public Number convertUnsigned(Number value) {
-    return scaleMissingUnsignedProxy.convertUnsigned(value);
+    return unsignedConversion != null ? unsignedConversion.convertUnsigned(value) : value;
+  }
+
+  public Number convertUnsigned(Number value, DataType dataType) {
+    return unsignedConversion != null ? unsignedConversion.convertUnsigned(value, dataType) : value;
```

Comment:
Not sure why you need this overload-- the `convertUnsigned(value, dataType)` being called here is the `static` version, maybe this is a mistake?

Reply:
yeah this is kinda a complicated design, I'm not in love with it. Basically we need the instance method (which is used above) to convert the variable data itself, which will get the datatype from the class instance. But the static method is needed for other properties on the variable like scale and offset, which can be a different datatype than the var, and that needs to be passed in directly.

Reply:
Okay, but since it's static maybe then we can do 
```suggestion
    return unsignedConversion != null ? UnsignedConversion.convertUnsigned(value, dataType) : value;
```